keps/sig-etcd/4578-feature-gate/README.md

```diff
@@ -0,0 +1,437 @@
+
+* Use feature gate as a mechanism to fix bugs.
+
+## Proposal
+
+### User Stories (Optional)
+
+#### Story 1
+
+It would be easier for developers to add new features without worrying too much about breaking etcd, because the feature would be Alpha first and disabled by default, and new changes would be mostly gated behind the feature gate.
+
+It would be a smoother process to move the feature to the next stage without the need to introduce a potentially breaking change of an `--experimental` flag.
+
+The developer also would not need to worry about feature compatibility when there is a mixed version cluster.
+
+#### Story 2
+
+Users of etcd cluster would be able to find out which features are enabled in the etcd cluster and decide on how to use them downstream. 
+
+Currently Kubernetes uses [`FeatureSupportChecker`](https://github.com/kubernetes/kubernetes/blob/db82fd1604ebf327ab74cde0a7158a8d95d46202/staging/src/k8s.io/apiserver/pkg/storage/etcd3/feature/feature_support_checker.go#L42) to check if a feature is supported by etcd, which basically checks the etcd version and compares the version with a hard-coded map of feature availability for different etcd versions.
+
+#### Story 3
+
+In a HA cluster, users should be able to enable/disable a feature without bring down the whole cluster, by restarting the servers with the new feature gate flag one by one. 
+
+#### Story 4
+
+During cluster upgrade/downgrade, feature changes across versions should have predictable behavior in this mix version scenario. 
+
+### Notes/Constraints/Caveats (Optional)
+
+#### Is feature enablement state a server level or cluster level property?
+
+There are features like `ExperimentalEnableLeaseCheckpoint` (enables leader to send regular checkpoints to other members to prevent reset of remaining TTL on leader change), if different server nodes have different enablement values for `ExperimentalEnableLeaseCheckpoint`, the results would be inconsistent depending on which one is the leader and confusing.
+
+There are also features like `ExperimentalEnableDistributedTracing` (enables distributed tracing using OpenTelemetry protocol), which could work fine at the local server level.
+
+We are proposing to have separate APIs to query if a feature is enabled for the local server, and for the whole cluster.
+
+#### Should we use feature gate for bug fixes?
+
+There are some use cases like `ProgressNotify`, for which some bugs are found and later fixed in a patch version. The client would need to know if the etcd version in use contains that fix to decide whether or not to use that feature. 
+
+The question is: should a new feature gate be added to signal the bug fix? 
+
+We think the answer is NO:
+* the new feature would need to be enabled by default to always apply the bug fix for new releases.
+* it changes the API which is not desirable in patch version releases.
```

Comment:
One exception to consider here is - a long-standing bug that can only be fixed by making a breaking API/semantic change. In some cases, the bug fix may be important enough to release in a patch version. To do that safely though, we'd likely need a feature gate. So we might wanna keep this door open.

A concrete example coming to my mind for the above is https://github.com/kubernetes/kubernetes/pull/123935.

Reply:
I do realize this is more of an exception than the norm. Nonetheless, it feels prudent for us to keep that door open.